synapseclient/models/link.py

```diff
@@ -0,0 +1,423 @@
+            "linksTo": {
+                "targetId": self.target_id,
+                "targetVersionNumber": self.target_version_number,
+            }
+            if self.target_id
+            else None,
+            "linksToClassName": self.links_to_class_name,
+        }
+        if request_dict["linksTo"]:
+            delete_none_keys(request_dict["linksTo"])
+        delete_none_keys(request_dict)
+        return request_dict
+
+    @otel_trace_method(
+        method_to_trace_name=lambda self, **kwargs: f"Link_Get: {self.id}"
+    )
+    async def get_async(
+        self,
+        parent: Optional[Union["Folder", "Project"]] = None,
+        follow_link: bool = False,
+        *,
+        synapse_client: Optional[Synapse] = None,
+    ) -> Union[
+        "Dataset",
+        "DatasetCollection",
+        "EntityView",
+        "File",
+        "Folder",
+        "MaterializedView",
+        "Project",
+        "SubmissionView",
+        "Table",
+        "VirtualTable",
+        "Link",
+    ]:
+        """Get the link metadata from Synapse. You are able to find a link by
+        either the id or the name and parent_id.
+
+        Arguments:
+            parent: The parent folder or project this link exists under.
+            follow_link: If True then the entity this link points to will be fetched
+                and returned instead of the Link entity itself.
+            synapse_client: If not passed in and caching was not disabled by
+                `Synapse.allow_client_caching(False)` this will use the last created
+                instance from the Synapse class constructor.
+
+        Returns:
+            The link object.
+
+        Raises:
+            ValueError: If the link does not have an id or a
+                (name and (`parent_id` or parent with an id)) set.
+        """
+        parent_id = parent.id if parent else self.parent_id
+        if not (self.id or (self.name and parent_id)):
+            raise ValueError(
+                "The link must have an id or a "
+                "(name and (`parent_id` or parent with an id)) set."
+            )
+        self.parent_id = parent_id
+
+        entity_id = await get_id(entity=self, synapse_client=synapse_client)
+
+        await get_from_entity_factory(
+            synapse_id_or_path=entity_id,
+            entity_to_update=self,
+            synapse_client=synapse_client,
+        )
+        self._set_last_persistent_instance()
+
+        if follow_link:
+            from synapseclient.models import FileOptions
+            from synapseclient.models.factory_operations import (
+                get_async as factory_get_async,
+            )
+
+            return await factory_get_async(
+                synapse_id=self.target_id,
+                version_number=self.target_version_number,
+                file_options=FileOptions(download_file=False),
+                synapse_client=synapse_client,
+            )
+        else:
+            return self
```

Comment:
If we get feedback or if reviewers have suggestions here I am more than happy to make changes.

What this `follow_link` logic does is to return the Entity that the link is actually pointing at. So if it's a file you'll get back the metadata about that FileEntity, one thing I was not certain on was if we should default behaivor around `download_file`.

Something we could possibly do here is to allow any of the options that the new `get/get_async` factory operation allows and pass it along from here.

Thoughts?

Reply:
This is defaulted to following the link to get the entity the Link points to, and one can set that behavior to False in order to get the Link itself